### PR TITLE
Add Smart Payment Buttons mode as alternative to directly embedded image links

### PR DESCRIPTION
**Pull request by dechov**
screen shot 2018-06-01 at 4 47 19 pm

(Toggled off by default for upgrading merchants, and on for new installs.)

Front-end view | Relevant settings | Fixes
-- | -- | --
Cart: screen shot 2018-06-01 at 9 57 57 am https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/99305a6524c65dc77f995febfd58dcfc18f4de7c | screen shot 2018-06-01 at 9 56 55 am https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/ea578d97abe0a2a41b3d7c910d932cdaf8e225c9 https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/2ca813a3256ccce9ec79a824b5b493992925750f https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/2318eb17eff01078a152a5ad734c7d821b6f975f https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/a9f4dec8b207e11548f377931ede72af0294190e
Single Product: screen shot 2018-06-01 at 9 10 31 am https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/d3727d6282b90a61c3aa578305a9406bbf9d619d | screen shot 2018-06-01 at 9 55 10 am https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/428ab6eb2a6ac34a6f624c91e1ae30b6622ff705 | https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/issues/407
Checkout: screen shot 2018-06-01 at 10 06 08 am https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/baf4f69cd29bebac51d95bc5445559d88f0f1ddd | screen shot 2018-06-01 at 10 07 09 am https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/b1137aa36469ecceb15c8df0ce9bc698ddb37d87 | https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/issues/403
Mini-cart: screen shot 2018-06-01 at 10 09 32 am https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/7ac071be3a4096a37f77368e5bda44b8e9dda90a | screen shot 2018-06-01 at 10 10 04 am https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/019648e37a8169311b57bdd8fedd2068a3d971b1 | https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/issues/328

### CODE REVIEW

**Comment by dechov** [comment]
Note: in the existing JS code for the single product page, a `generate_cart` call is made before redirecting to PayPal. With the new integration, this could _probably_ just as well be done during the `start_checkout` call, but I've left it as a separate call preceding it so as to reduce the potential for unforeseen side effects. Open to changing it.

Another thing I should note: although little or no code is actually removed in this PR, a future update will remove the legacy mode, and though it's not yet precisely clear what we can remove at that point, there will likely be a large amount to clean up.

File: includes/class-wc-gateway-ppec-cart-handler.php [hunk 1]
```diff
@@ -223,11 +299,68 @@ public function enqueue_scripts() {
...
+				} else {
+					$page_data = array(
+						'button_layout'        => $settings->button_layout,
+						'button_size'          => $settings->button_size,
+						'hide_funding_methods' => $settings->hide_funding_methods,
+						'credit_enabled'       => $settings->credit_enabled,
+					);
+				}
+
+				$data = array_merge( $data, $this->convert_to_render_settings( $page_data ) );
```
Comment (jeffstieler) [inline includes/class-wc-gateway-ppec-cart-handler.php]: Since these 4 settings are repeated a lot, and are the only ones inspected/modified in `convert_to_render_settings()` - what do you think about combining the retrieval of them with an optional prefix from `$settings` and the "conversion" into a single method? Something like `get_button_settings( $settings, 'single_product' )`?

File: includes/class-wc-gateway-ppec-with-spb.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+	/**
+	 * Process payments.
+	 *
+	 * @param int $order_id Order ID
+	 *
+	 * @return array
+	 */
+	public function process_payment( $order_id ) {
+		if ( isset( $_POST['payerID'] ) ) {
```
Comment (jeffstieler) [inline includes/class-wc-gateway-ppec-with-spb.php]: We should probably check for `paymentToken` here too.

File: includes/class-wc-gateway-ppec-with-spb.php [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+		<div id="woo_pp_ec_button"></div>
+		<?php
+	}
+
+	/**
+	 * Outputs scripts used for PPEC payment.
+	 *
+	 * @since 1.6.0
+	 */
+	public function payment_scripts() {
```
Comment (jeffstieler) [inline includes/class-wc-gateway-ppec-with-spb.php]: Can you document this behavior with a comment?

File: includes/class-wc-gateway-ppec-with-spb.php [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	}
+
+	/**
+	 * Process payments.
+	 *
+	 * @param int $order_id Order ID
+	 *
+	 * @return array
+	 */
+	public function process_payment( $order_id ) {
```
Comment (jeffstieler) [inline includes/class-wc-gateway-ppec-with-spb.php]: Can you document this behavior with a comment?

File: includes/settings/settings-ppec.php [hunk 5]
```diff
@@ -107,13 +107,115 @@
...
+			button_size.find( 'option[value=\"responsive\"]' ).prop( 'disabled', ! checked );
+			button_size.find( 'option[value=\"small\"]' ).prop( 'disabled', false );
+			$( document.body ).trigger( 'wc-enhanced-select-init' );
+
+			if ( checked ) {
+				$( '.woocommerce_ppec_paypal_button_size' ).val( 'responsive' ).change();
+			} else if ( ! button_size.val() ) {
+				button_size.val( 'large' ).change();
+			}
+		} ).change();
```
Comment (jeffstieler) [inline includes/settings/settings-ppec.php]: All of this conditional form input logic needs documenting.

File: includes/settings/settings-ppec.php [hunk 6]
```diff
@@ -400,4 +457,168 @@
...
 	),
-) );
+
+	'button_settings' => array(
+		'title'       => __( 'Button Settings', 'woocommerce-gateway-paypal-express-checkout' ),
+		'type'        => 'title',
+		'description' => __( 'Customize the appearance of Express Checkout on your site.', 'woocommerce-gateway-paypal-express-checkout' ),
+	),
+	'use_spb' => array(
+		'title'       => __( 'Smart Payment Buttons', 'woocommerce-gateway-paypal-express-checkout' ),
```
Comment (jeffstieler) [inline includes/settings/settings-ppec.php]: Duplicate key.

File: includes/settings/settings-ppec.php [hunk 7]
```diff
@@ -400,4 +457,168 @@
...
+$settings['mark_settings_toggle'] = array(
+	'title'   => __( 'Configure settings specific to regular checkout', 'woocommerce-gateway-paypal-express-checkout' ),
+	'type'    => 'checkbox',
+	'class'       => 'woocommerce_ppec_paypal_spb woocommerce_ppec_paypal_visibility_toggle',
+	'default' => 'no',
+);
+foreach( $per_context_settings as $key => $value ) {
+	$value['class'] .= ' woocommerce_ppec_paypal_mark';
+	$settings[ 'mark_' . $key ] = $value;
+}
```
Comment (jeffstieler) [inline includes/settings/settings-ppec.php]: A comment breaking up each "context" would make this much more readable.

File: assets/js/wc-gateway-ppec-smart-payment-buttons.js [hunk 8]
```diff
@@ -0,0 +1,99 @@
+/* global wc_ppec_context */
+;( function ( $, window, document ) {
+	'use strict';
+
+	var getFundingMethods = function( methods ) {
```
[the remark this answers is not in the document]
Comment (dechov) [reply assets/js/wc-gateway-ppec-smart-payment-buttons.js]: True, but `_` isn't available here – would you suggest loading it?

File: includes/settings/settings-ppec.php (hunk 6 above)
Comment (dechov) [reply includes/settings/settings-ppec.php]: Thanks – not sure when or how the latter one got changed from `label` to `title`.

File: includes/class-wc-gateway-ppec-cart-handler.php [hunk 9]
```diff
@@ -118,6 +119,24 @@ public function wc_ajax_update_shipping_costs() {
...
+	public function wc_ajax_start_checkout() {
+		if ( ! wp_verify_nonce( $_POST['nonce'], '_wc_ppec_start_checkout_nonce' ) ) {
+			wp_die( __( 'Cheatin&#8217; huh?', 'woocommerce-gateway-paypal-express-checkout' ) );
+		}
+
+		if ( 'yes' === $_POST['from_checkout'] ) {
+			add_filter( 'woocommerce_cart_needs_shipping', '__return_false' );
+		}
+
+		wc_gateway_ppec()->checkout->start_checkout_from_cart();
```
[the remark this answers is not in the document]
Comment (dechov) [reply includes/class-wc-gateway-ppec-cart-handler.php]: Good question – will test and make sure errors are handled and reported.

File: includes/class-wc-gateway-ppec-cart-handler.php (hunk 1 above)
Comment (dechov) [reply includes/class-wc-gateway-ppec-cart-handler.php]: Done in https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/86932eefb55ab497b11d51c4714b36a3aa58327d

File: assets/js/wc-gateway-ppec-smart-payment-buttons.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (dechov) [reply assets/js/wc-gateway-ppec-smart-payment-buttons.js]: Did the latter in https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/9121cc64dc3f2394bb4221b9cb77f1c756f68e5b

**Comment by dechov** [comment]
Added comments and addressed other feedback. Let me know if there's any more code that's difficult to read or understand.

Also enabled SPB mode by default for new installs, and added an admin notice when enabled but not in SPB mode:

screen shot 2018-06-18 at 10 20 20 pm

Still need to add descriptions but opening back up to code review in the meantime.

File: includes/settings/settings-ppec.php [hunk 10]
```diff
@@ -499,7 +499,7 @@ function showHideDefaultButtonSettings() {
 	'use_spb' => array(
 		'title'       => __( 'Smart Payment Buttons', 'woocommerce-gateway-paypal-express-checkout' ),
 		'type'        => 'checkbox',
-		'default'     => 'no',
+		'default'     => $this->get_option( 'button_size' ) ? 'no' : 'yes', // Hinge on whether settings have been initialized.
```
Comment (jeffstieler) [inline includes/settings/settings-ppec.php]: Nit: being more specific in this comment would be helpful. E.g. `button_size` being set indicates an existing PPEC install.

File: includes/class-wc-gateway-ppec-plugin.php [hunk 11]
```diff
@@ -222,6 +226,29 @@ public function show_bootstrap_warning() {
...
+			$setting_link = $this->get_admin_setting_link();
+			$message = sprintf( __( '<p>PayPal Checkout with new <strong>Smart Payment Buttons™</strong> give your customers the power to pay the way they want without leaving your site.</p><p>The <strong>existing buttons will be deprecated and removed</strong> in future releases. Upgrade to Smart Payment Buttons in the <a href="%s">settings</a>.</p>', 'woocommerce-gateway-paypal-express-checkout' ), esc_url( $setting_link ) );
+			?>
+			<div class="notice notice-warning is-dismissible ppec-dismiss-spb-notice">
+				<?php echo wp_kses( $message, array( 'a' => array( 'href' => array() ), 'strong' => array(), 'p' => array() ) ); ?>
+			</div>
+			<script>
+			( function( $ ) {
+				$( '.ppec-dismiss-spb-notice' ).on( 'click', '.notice-dismiss', function() {
+					jQuery.post( "<?php echo admin_url( 'admin-ajax.php' ); ?>", {
```
[the remark this answers is not in the document]
Comment (dechov) [reply includes/class-wc-gateway-ppec-plugin.php]: Done in https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/5113019867a9e4b744aeb1645e5e4a5d244bf9c8

File: includes/class-wc-gateway-ppec-plugin.php [hunk 12]
```diff
@@ -227,9 +225,22 @@ public function show_bootstrap_warning() {
...
+
+		// Should only show on WooCommerce screens, the main dashboard, and on the plugins screen (as in WC_Admin_Notices).
+		if ( ! in_array( $screen_id, wc_get_screen_ids(), true ) && 'dashboard' !== $screen_id && 'plugins' !== $screen_id ) {
+			return;
+		}
+
+		// Should only show when PPEC is enabled but not in SPB mode.
+		if ( 'yes' !== $this->settings->enabled || 'yes' === $this->settings->use_spb ) {
+			return;
+		}
```
[the remark this answers is not in the document]
Comment (dechov) [reply includes/class-wc-gateway-ppec-plugin.php]: Done in https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/9a8e6e569846e24ba9baeb1f8b6194f6cf8426ba

File: includes/settings/settings-ppec.php (hunk 10 above)
Comment (dechov) [reply includes/settings/settings-ppec.php]: Done in https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/383872cdbdb793ec8d7edca591aeb5a851385fba

File: includes/settings/settings-ppec.php [hunk 13]
```diff
@@ -521,6 +523,7 @@ function showHideDefaultButtonSettings() {
 		'class'       => 'wc-enhanced-select woocommerce_ppec_paypal_spb',
 		'default'     => 'rect',
 		'desc_tip'    => true,
+		'description' => __( 'The pill-shaped button\'s unique and powerful shape signifies PayPal in people’s minds. Use the rectangular button as an alternative when pill-shaped buttons might pose design challenges.', 'woocommerce-gateway-paypal-express-checkout' ),
```
Comment (jeffstieler) [inline includes/settings/settings-ppec.php]: Intentional usage of both `'` and `’` ?

Comment (dechov) [reply includes/settings/settings-ppec.php]: Nice catch – changed in https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/088b59cff622190aa7a2f413b5e55b579c2348a8.

**Comment by othqsa95** [comment]
Hello everyone,

I added the PAYPAL button in the MINI-CART and I adjusted it well with the CSS. I note however, that when I add a product to the basket, the slide cart opens and the button is displayed correctly.

But when I leave the slide cart and open it a second time without adding new products, the Paypal button is not displayed.

Anyone have an idea to solve this little problem?
